### ISSUE TITLE
Anaconda Python ImportError: No module named 'mkdocs'

### ISSUE THREAD

**Issue by ethen8181** [mkdocs/mkdocs#1007]
Hi, I'm getting `ImportError: No module named 'mkdocs'` after calling `mkdocs build` as shown below.

<img width="680" alt="screen shot 2016-08-01 at 4 21 22 pm" src="https://cloud.githubusercontent.com/assets/12273134/17287757/107b5cb0-5804-11e6-9e57-9c969974ccff.png">

I've tried upgrading pip and uninstalling and reinstalling mkdocs but still obtained the same error. Any idea of how may I fixed this? Thanks.

pip and mkdocs's version are printed below.

<img width="636" alt="screen shot 2016-08-01 at 4 21 14 pm" src="https://cloud.githubusercontent.com/assets/12273134/17287759/107cf976-5804-11e6-981d-d01556c3cf2b.png">

**Comment by waylan** [comment]
Your screenshots don't seem to match what you are saying. The `build` command worked fine. In the first screenshot, the first two commands work correctly as do the commands in the second second shot. The only issue is with the third command in the first screenshot, the same command which worked only two commands earlier.

I'm not sure, but I would guess this is not a MkDocs issue but an issue with how Python is installed on your system. I note that your error messages indicate that the Python libs are installed under Anaconda, which suggests to me that you are not using a System wide install of Python. I'm not real familiar with Anaconda, so I don't know, but it may even be using virtual environments. Perhaps between the second and third command you disabled/switched virtualenvs from within Anaconda (or opened/closed Anaconda). In any event, this definitely smells of multiple Pythons installed and you not using the same one each time (in other words system misconfiguration). Try running `which python` both when the problem exists and when it does not and I expect you will find a different Python is being used.

**Comment by ethen8181** [comment]
Hi, Thanks for the fast response and sorry for the confusion. What I meant was after calling `mkdocs build` I get ImportError: No module named 'mkdocs'.

I've tried what you've said, but I don't think a different python is being used (screenshots below).

<img width="513" alt="screen shot 2016-08-01 at 9 35 08 pm" src="https://cloud.githubusercontent.com/assets/12273134/17295673/dceb9a96-582f-11e6-96ea-817d0df7c649.png">

**Comment by waylan** [comment]
Try running the command as a sub-command of Python:

    python -m mkdocs

Does that make a difference?

**Comment by ethen8181** [comment]
Like this?

<img width="681" alt="screen shot 2016-08-01 at 9 53 34 pm" src="https://cloud.githubusercontent.com/assets/12273134/17296188/7adf5d30-5832-11e6-8a0c-8e0aaef168b1.png">

**Comment by waylan** [comment]
I'm out of ideas.  It seems clear to me that there is some weird configuration issue on your system. As Python is installed under Anaconda, I would suggest checking with them to see if this is a known problem.

**Comment by waylan** [comment]
One other thing you might want to check is whether this error occurs with other installed Python scripts? Of course, which ones would depend on what you have installed. Based on your error messages, it would appear that the `mkdocs` script is at `/Users/ethen/anaconda/bin/`, so any other python scripts in that directory would be good candidates.

**Comment by ethen8181** [comment]
Everything else that I usually use seems to be working. I'll just paste the conda version here in case someone else happens to have the same issue in the future .. Thanks for the help!

<img width="559" alt="screen shot 2016-08-02 at 8 52 23 am" src="https://cloud.githubusercontent.com/assets/12273134/17313502/c0c836f8-588e-11e6-844a-3e13d390a81d.png">

**Comment by emorfam** [comment]
I can confirm this issue - I have the same problem with Anaconda. That's the first time I have this problem with a Python package that is managed with Anaconda, so I don't believe Anaconda is responsible. Is it possible that this problem is due to the fact that Anaconda doesn't use the PYTHONPATH variable but PATH instead?

**Comment by waylan** [comment]
We're just using standard Python/setuptools stuff for installation. No custom funny business. Not sure how a different environment variable would make a difference (if it works fine with other packages), although if Anaconda is using non-standard stuff, then my inclination is not not care if it doesn't work and suggest you use standard Python.

In any event, I can't reproduce this outside of Anaconda and I don't have Anaconda installed (nor do I have the desire to install it). However, if someone can provide a patch I'll happily merge it.

**Comment by pivaldi** [comment]
Same problem here in Debian 9.0 (stretch), I installed mkdocs via pip.
All is right with the command ` python -m mkdocs serve`.

```txt
$ python --version
Python 2.7.12+
```

In */usr/local/bin/mkdocs*, replacing the shebang `#!/usr/bin/python3` by `#!/usr/bin/env python`, fixe the problem.

**Comment by waylan** [comment]
> In */usr/local/bin/mkdocs*, replacing the shebang `#!/usr/bin/python3` by `#!/usr/bin/env python`, fixe the problem.

That sounds like a bug in setuptools, which generates that file for us. Not sure what we can do about it.

**Comment by waylan** [comment]
I've updated the title to reflect the fact that this issue is specific to Anaconda Python. I recently came across [this][1], which explains why there are differences with Anaconda installations:

> conda is the package management tool for [Anaconda][2] Python installations. Anaconda Python is a distribution from [Continuum Analytics][3] specifically aimed at the scientific community, and in particular on Windows where the installation of binary extensions is often difficult.
> 
> Conda is a completely separate tool to pip, virtualenv and wheel, but provides many of their combined features in terms of package management, virtual environment management and deployment of binary extensions.
> 
> Conda does not install packages from PyPI and can install only from the official Continuum repositories, or anaconda.org (a place for user-contributed conda packages), or a local (e.g. intranet) package server. However, note that pip can be installed into, and work side-by-side with conda for managing distributions from PyPI.

Given the above, either we need to test in Anaconda Python or explicitly state we do not support it.

[1]: https://packaging.python.org/key_projects/#conda
[2]: http://docs.continuum.io/anaconda/index.html
[3]: http://continuum.io/downloads

**Comment by tramponi** [comment]
I have no problems at all with mkdocs 0.16.3, conda 4.3.24  and Python 3.6.2 on Windows (x64).

![conda](https://user-images.githubusercontent.com/18346173/29459204-e3830cc0-8422-11e7-9eab-8a5b7327c9da.png)

**Comment by nathancarter** [comment]
I am having this same problem on my Mac, with Python 2.7 installed in /Library/Python/2.7.  I can give more specifics.

If I run mkdocs in a folder in which it has not yet been run, then it succeeds.  This includes commands such as `mkdocs`, `mkdocs --version`, `mkdocs build`, `mkdocs serve`, and `mkdocs new`.

If I run mkdocs in a folder in which it has already been run, and thus a `site/` folder exists, then it fails.  This includes `mkdocs`, `mkdocs --version`, `mkdocs build`, `mkdocs serve`, and `mkdocs new` all failing with the same error.
```
Traceback (most recent call last):
  File "/usr/local/bin/mkdocs", line 7, in <module>
    from mkdocs.__main__ import cli
ImportError: No module named mkdocs.__main__
```

I recently uninstalled mkdocs through pip, then re-installed it, so I'm running version 0.16.3, a fresh install.

**Comment by nathancarter** [comment]
Also, I don't know what Anaconda is, so I don't think this issue is specific to that (unless I just don't know I'm using it??).  And my current workaround is to simply `rm -rf site/` and then run `mkdocs build` and it rebuilds fine.  This is obviously not ideal.

**Comment by waylan** [comment]
@nathancarter Just curious, does the problem persist if you change the [`site_dir`][1] in the config to something other than `site`?

Also, if the problem exists when the "site" dir exists, but not when it doesn't (btw, no previous reporter specifically mentioned that detail; however, on review, it appears that that is the case for each reporter) I find that highly suspicious. Where is your  `site_dir` located in relation to the `config.yml` file and `docs_dir`? And what are the contents of the directory (a list of files and directories)? I have some suspicions about what is going on, but I need that info to know for sure. Or you could point me to a publicly available copy of your project if one exists.

[1]: http://www.mkdocs.org/user-guide/configuration/#site_dir

**Comment by nathancarter** [comment]
Correction:  Since I've installed the latest mkdocs through pip, the problem has gone away.  I did not realize it at first, because I was juggling too many terminal windows.  My apologies for the incorrect bug report up above.

**Comment by waylan** [comment]
Taking another look at this issue, I am certain this is a symptom of #807, which was fixed in [version 0.16][1] (see ac2ff10afd). And all previous reports of this problem were made before 0.16 was released.  With the confirmation from @nathancarter that updating to 0.16+ resolves the problem on a system on which it existed, I'm closing this as a duplicate of #807.

By way of explanation, my assumption is that because some Python files (at least `__init__.py`) were being copied into the `site_dir`, Python considered the `site_dir` to be a Python module on the PYTHONPATH. Therefore, when running MkDocs, some import would fail because it was trying to import the `site_dir` rather than the appropriate library (probably because the `site_dir` was in the `cwd`). However, different systems (even different Python installs on the same system) can be configured to resolve PYTHONPATH differently, so it was difficult to reproduce by others. And when you can't reproduce it, it's really hard to debug. In the end, it appears that Anaconda was a red herring. While it is apparently one of the systems who's default configuration in such that the problem arises, it would by no means be the only system which could have that configuration. Presumably, now that we know the cause, the problem could be reproduced (with MkDocs version 0.15.x) on any system with a few minor adjustments to how PYTHONPATH is resolved. Regardless, the correct fix has already been made in 0.16, which ensures that the `site_dir` cannot be mistaken for a Python module regardless of how Python is configured.

[1]: http://www.mkdocs.org/about/release-notes/#version-016-2016-11-04
